### PR TITLE
Initial isotherm plugin

### PR DESCRIPTION
**Pull request by bobmyhill**
This PR adds an Initial temperature plugin which allows the user to specify the depths of isotherms as a function of "horizontal" coordinates (x, y or phi, theta) in a set of ASCII files. The depth of the isotherms are interpolated linearly as a function of horizontal natural coordinates (either x-y or lat-lon), and then the temperature structure is interpolated vertically as piecewise constant or linear between isotherms.

The plugin uses AsciiDataLayered, a new class in Utilities. This could be used for a variety of different plugins (for example, to parameterize Crust 1.0).

### CODE REVIEW

File: include/aspect/initial_temperature/ascii_data_layered.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+         * beginning of the program. Checks preconditions.
+         */
+        void
+        initialize ();
+
+        // avoid -Woverloaded-virtual:
+        using Utilities::AsciiDataLayered<dim>::initialize;
+
+        /**
+         * Return the boundary temperature as a function of position. For the
```
Comment (tjhei) [inline include/aspect/initial_temperature/ascii_data_layered.h]: copy and pasted?

File: include/aspect/initial_temperature/ascii_data_layered.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    using namespace dealii;
+
+    /**
+     * A class that implements a prescribed temperature field determined from
+     * a AsciiData input file.
```
Comment (tjhei) [inline include/aspect/initial_temperature/ascii_data_layered.h]: please update

File: source/initial_temperature/ascii_data_layered.cc [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+#include <aspect/global.h>
+#include <aspect/initial_temperature/ascii_data_layered.h>
+
+namespace aspect
+{
+  namespace InitialTemperature
+  {
+    template <int dim>
+    AsciiDataLayered<dim>::AsciiDataLayered ()
+    {}
```
Comment (tjhei) [inline source/initial_temperature/ascii_data_layered.cc]: can the constructor be deleted?

File: source/utilities.cc [hunk 4]
```diff
@@ -2277,6 +2277,182 @@ namespace aspect
 
 
 
+    template <int dim>
+    AsciiDataLayered<dim>::AsciiDataLayered ()
+    {}
```
Comment (tjhei) [inline source/utilities.cc]: remove?

File: source/utilities.cc [hunk 5]
```diff
@@ -2277,6 +2277,182 @@ namespace aspect
...
+    template <int dim>
+    void
+    AsciiDataLayered<dim>::initialize(const unsigned int components)
+    {
+      AssertThrow ((dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model()))
+                   || (dynamic_cast<const GeometryModel::Chunk<dim>*> (&this->get_geometry_model())) != nullptr
+                   || (dynamic_cast<const GeometryModel::Sphere<dim>*> (&this->get_geometry_model())) != nullptr
+                   || (dynamic_cast<const GeometryModel::Box<dim>*> (&this->get_geometry_model())) != nullptr,
+                   ExcMessage ("This ascii data plugin can only be used when using "
+                               "a spherical shell, chunk or box geometry."));
```
Comment (tjhei) [inline source/utilities.cc]: or sphere

File: source/utilities.cc [hunk 6]
```diff
@@ -2277,6 +2277,182 @@ namespace aspect
...
+
+      // Create the lookups for each file and a vector of doubles corresponding to the layer_values
+      std::string filename;
+      for (unsigned int i=0; i<layer_boundary_names.size(); ++i)
+        {
+          filename = data_directory + data_file_names[i];
```
Comment (tjhei) [inline source/utilities.cc]: const std::string filename =

File: data/initial-temperature/ascii-data/test/initial_isotherm_1000K_box_2d.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Test data for ascii data initial temperature conditions.
```
Comment (tjhei) [inline data/initial-temperature/ascii-data/test/initial_isotherm_1000K_box_2d.txt]: you need to update the comments on all of these files.

File: source/utilities.cc [hunk 8]
```diff
@@ -2277,6 +2277,182 @@ namespace aspect
...
+                                             const std::string &subsection_name)
+    {
+      Utilities::AsciiDataBase<dim>::parse_parameters(prm, subsection_name);
+
+      prm.enter_subsection(subsection_name);
+      {
+
+        data_directory = Utilities::expand_ASPECT_SOURCE_DIR(prm.get ("Data directory"));
+        data_file_names    = Utilities::split_string_list(prm.get ("Data file names"), ',');
+        layer_boundary_names = Utilities::split_string_list(prm.get("Layer boundary values"), ',');
```
Comment (tjhei) [inline source/utilities.cc]: are those names or values?!

Comment (bobmyhill) [reply source/utilities.cc]: Ooops, I used this parameter when I had a different file-finding structure. Changed!

File: source/initial_temperature/ascii_data_layered.cc (hunk 3 above)
Comment (bobmyhill) [reply source/initial_temperature/ascii_data_layered.cc]: I don't think so? At least, when I try, I get a bunch of `error: type ‘aspect::Utilities::AsciiDataLayered<2>’ is not a base type for type ‘aspect::InitialTemperature::AsciiDataLayered<2>’` errors

**Comment by bobmyhill** [comment]
> I am not quite sure what the difference between this and the existing asciidata plugin is. Maybe this because some of the docs are missing? Can you explain (or even better extend the comments)?

In the other AsciiData plugin, the temperature is read in using a <dim>D spatial grid that spans the domain. In this plugin, the data is read in as a series of <dim-1>D grids, which each represent an isotherm. For each point in the domain, the temperature is evaluated by first finding the isotherms bounding the data point, then finding the depth of those isotherms, and finally interpolating between them (either linearly or piece-wise constant). 

This is a much more compact data structure than the full <dim>D grid for many geological applications (because vertical/radial temperature profiles are often well-approximated by a few linear T-z segments).

**Comment by bobmyhill** [comment]
I just noticed that if we want to use this for Crust 1 and other similar composition initial conditions, we need to allow AsciiDataLayered to interpolate properties on the layer boundaries. The new commit implements that change.

File: include/aspect/utilities.h [hunk 9]
```diff
@@ -941,6 +941,85 @@ namespace aspect
...
+        void
+        initialize (const unsigned int components);
+
+
+        /**
+         * Returns the data component at the given position.
+         */
+        double
+        get_data_component (const Point<dim>                    &position,
+                            const unsigned int                   component) const;
```
Comment (gassmoeller) [inline include/aspect/utilities.h]: Please remove the many spaces between the parameters and their types.

File: source/initial_temperature/ascii_data_layered.cc [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+
+    template <int dim>
+    void
+    AsciiDataLayered<dim>::initialize ()
+    {
+      Utilities::AsciiDataLayered<dim>::initialize(2);
```
Comment (gassmoeller) [inline source/initial_temperature/ascii_data_layered.cc]: Hmm, why 2? Should it not be a single temperature component?

File: source/utilities.cc [hunk 11]
```diff
@@ -2277,6 +2277,174 @@ namespace aspect
...
+
+
+    template <int dim>
+    void
+    AsciiDataLayered<dim>::initialize(const unsigned int components)
+    {
+      AssertThrow ((dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model()))
+                   || (dynamic_cast<const GeometryModel::Chunk<dim>*> (&this->get_geometry_model())) != nullptr
+                   || (dynamic_cast<const GeometryModel::Sphere<dim>*> (&this->get_geometry_model())) != nullptr
+                   || (dynamic_cast<const GeometryModel::Box<dim>*> (&this->get_geometry_model())) != nullptr,
```
Comment (gassmoeller) [inline source/utilities.cc]: I know this is still common in other parts of the code, but could you please replace this with the new and easier structure: `Plugins::plugin_type_matches<GeometryModel::Box<dim> >(this->get_geometry_model())` ? Same for all the other checks.

File: source/utilities.cc [hunk 12]
```diff
@@ -2277,6 +2277,174 @@ namespace aspect
...
+        }
+    }
+
+
+
+    template <int dim>
+    double
+    AsciiDataLayered<dim>::
+    get_data_component (const Point<dim>                    &position,
+                        const unsigned int                   component) const
```
Comment (gassmoeller) [inline source/utilities.cc]: please remove spaces

File: source/utilities.cc [hunk 13]
```diff
@@ -2277,6 +2277,174 @@ namespace aspect
...
+        }
+
+      // Find which layer we're in
+      unsigned int layer_boundary_index=0;
+      // if position < position of the first boundary layer, stop
+      double old_difference_in_vertical_position = vertical_position - lookups[layer_boundary_index]->get_data(horizontal_position,0);
+      double difference_in_vertical_position = old_difference_in_vertical_position;
+      while (difference_in_vertical_position > 0. && layer_boundary_index < number_of_layer_boundaries-1)
+        {
+          layer_boundary_index++;
```
Comment (gassmoeller) [inline source/utilities.cc]:
```suggestion
          ++layer_boundary_index;
```

File: source/utilities.cc [hunk 14]
```diff
@@ -2333,8 +2501,6 @@ namespace aspect
       return lookup->get_data(internal_position,component);
     }
 
-
-
```
Comment (gassmoeller) [inline source/utilities.cc]: Please leave these line in here.

**Comment by bobmyhill** [comment]
Thanks @gassmoeller , I've addressed your comments, so this should be good to go as and when the tester is done.

**Comment by gassmoeller** [comment]
Thanks!

**Pull request merged**